### ISSUE TITLE
the release of a click on the tab-title should put the content of the tab in the focus

### ISSUE THREAD

**Issue by ralf3u** [xiota/geany-preview#15]
actual behavior
If I click on a tab-title of the sidebar: The tab-title is in focus, even if the mouse-key is not pressed any more. Evidence: There is no sign for it, but both tests below are negative.

control tests with a very long html-document, and with at least two tabs in the sidebar:
TEST 1: Open Geany, click on the Preview-tab-title, then press the arrow-left-key only once: the content of the Preview-tab should still be in focus.
TEST 2: Open Geany, click on the Preview-tab-title, then press the arrow-down-key only once: the html-document should move downwards.

expectation
- The tab-title of Preview should be in focus as long as the mouse-key is pressed, but in the moment where the mouse-key is not pressed any more, the content of the Preview-tab should be in focus.
- It should work for all tabs of the sidebar.

remark
The tabs in the sidebar should work exactly like the tabs in the editor: Open Geany, click on a tab-title of the editor and hold the mouse-key for about 10 seconds: tab-title is in focus, but in the moment where the mouse-key is not pressed any more, the content of the tab is in focus, and not the tab-title of the editor, what is visible with the highlighted line.

**Comment by ralf3u** [comment]
Maybe the solution of this issue has a connection with #5.

**Comment by ralf3u** [comment]
The main problem of #5 is the detection of a signal whether sidebar is in focus or not in focus when clicking in the sidebar. This signal apparently does not exist in Geany.

But the problem of #15 is different: When clicking on the tab-title of the sidebar, the tab-title of the sidebar becomes green. I don't know how to program software, but does not something exist like:

`IF there is a CSS-color change of the tab-title, THEN execute something like geany.keycmd("FOCUS_SIDEBAR")`.

(For more information to `geany.keycmd("FOCUS_SIDEBAR")`, see https://github.com/xiota/geanylua-scripts/issues/7#issuecomment-1455335094).

**Comment by ralf3u** [comment]
> I don't know how to program software, but does not something exist like:
`IF there is a CSS-color change of the tab-title, THEN execute something like geany.keycmd("FOCUS_SIDEBAR").`

Is this possible or not possible?

**Comment by xiota** [comment]
Setting focus has been an ongoing issue that may not be resolvable any time soon.

**Comment by ralf3u** [comment]
In September 2022 I opened at Geany the issue number 3289. In that issue there are two links about GTK3 that could help to solve the focus issue:

https://docs.gtk.org/gtk3/signal.Notebook.focus-tab.html
https://docs.gtk.org/gtk3/signal.Widget.grab-focus.html

Did you try both?

**Comment by xiota** [comment]
I tried every signal a couple years ago.  None of them worked at the time.  The problem is focus goes where you don't want it.  You want the browser contents to be in focus.  But usually some other part of the notebook is given focus, like the tabbar.

**Comment by ralf3u** [comment]
OK, but what about the idea to report at GTK that some things don't work as they should?

I think that there are two issues that should be reported at GTK because they don't work correctly:
- to detect if content-area of the tab has focus,
- to put the focus on the content-area of the tab.

I will not report the issues because I have too less knowledge about programing software.

**Comment by xiota** [comment]
> I will not report the issues because I have too less knowledge about programing software.

I don't know enough to be sure there is a problem with GTK or whether it's working the way they want it to.  They may not want programmers to be able to move focus around in certain circumstances because it's a potential security risk (eg, to intercept user input).

**Comment by ralf3u** [comment]
>I don't know enough to be sure there is a problem with GTK or whether it's working the way they want it to.

I would say: Just report it and then you will know with the answer they offer.

**Comment by xiota** [comment]
> Just report it and then you will know with the answer they offer.

I don't really want to write code to reproducibly replicate the problem, try fixes, etc.  Plus, GTK development has moved on to GTK4.  I don't even know if Geany and plugins build with GTK4.

**Comment by ralf3u** [comment]
> I don't even know if Geany and plugins build with GTK4.

Maybe this information helps: https://github.com/xiota/geany-preview/issues/5#issuecomment-1261747888.

**Comment by xiota** [comment]
I doubt the devs will proactively update that issue.  When they do decide to support GTK4, the devs (not users) will probably have to open a lot of new issues to remove outdated GTK3 stuff.

**Comment by ralf3u** [comment]
So, that means for me that you wait till Geany und the plugins will support GTK4, and then, if the issue is still there, you would report it to GTK. The problem could be then that GTK5 will be already available. So, it could be that the issue will never be reported.

> When they do decide to support GTK4, the devs (not users) will probably have to open a lot of new issues to remove outdated GTK3 stuff.

That's your point of view. I have another point of view: If you report the issues now, then I can imagine that they will not fix it for GTK3 but maybe think about a solution that will be supported only by GTK4 and future versions.

**Comment by xiota** [comment]
I am not going to open bug reports to GTK about this.  The entire process will take a lot of time and effort.  Most likely nothing would come of it.

Why don't you figure out how to access and search their bug reports to see whether it's already been reported.  Then you can track it's process.

**Comment by ralf3u** [comment]
Well, I can try to report at GTK. Let's try first with the title. I would describe it like this:
issue 1:
GTK3: no possibility to put focus on the content-area of a tab of the sidebar in the editor Geany
issue 2:
GTK3: no possibility to detect if content-area of a tab of the sidebar is in focus in the editor Geany

Is it OK like this?

**Comment by xiota** [comment]
Do a search first to see if you find anything relevant.  I don't really want to be involved in an upstream bug report.

**Comment by ralf3u** [comment]
OK

**Comment by ralf3u** [comment]
I just recognized that in Lubuntu 23.10 with Geany 1.38 if I click in the sidebar on the tab-title `Files` (plugin `File Browser`), then the content of the tab is also not in focus. I will report it the next days at geany-plugins.

**Comment by ralf3u** [comment]
Before reporting the issue at geany-plugins, I wanted to ask if the information could help to solve the issue:
https://docs.gtk.org/gtk3/method.Widget.child_focus.html

**Comment by xiota** [comment]
I don't see how this is related to geany-plugins.

**Comment by ralf3u** [comment]
I just opened at geany-plugins an issue with the issue number 1320. The title of the issue is:
"A release of a click on the tab-title of a plugin-tab in the sidebar should put the content of the tab in focus".

**Comment by ralf3u** [comment]
The plugin Preview is not responsible for this issue. The issue has to be fixed upstream. That is the reason why I close this issue.

**Comment by ralf3u** [comment]
Could you please join the conversation at geany-plugins 1320, because the maintainer claims that the issue does not belong to the Geany project.

**Comment by ralf3u** [comment]
Now the maintainer added some more information:

>Therefore plugins are entirely responsible for the behaviour of their UI. Plugins should connect to the relevant signal eg "switch-page" or "focus-tab" or "clicked" on the tab or whatever and then implement the behaviour they require.

So, the sentences for me means that Preview is the problem. The sentences means also for me that it can be solved in the plugin Preview.

>[...] I said elsewhere the tab and the contents are different things, and in standard GTK the tab can be focussed instead of the content by clicking it when the page is current.

When I observed the behavior of the plugin `File Browser`, I thought that it was the same for the plugin Preview, namely that the toolbar is in focus. But now I think that it is not the toolbar that is in focus after a release of a click on the Preview-tab, it is the tab, and that is the reason why I have first to press 1x the arrow-down-key on the keyboard to put the content of the tab in focus.

**Comment by xiota** [comment]
When the tab is clicked, the tab itself is in focus.  Probably not a good idea to change focus to the contents because that would interfere with changing tabs with left/right.  You would then have to use ctrl+pgdn / ctrl+pgup.

Figuring out the exact component of interest and the right signals to use is difficult and unreliable.  Focus may not go where you want or expect with every sidebar plugin.  I already tried unsuccessfully to change focus to the gtkwebkit widget in the past.

**Comment by ralf3u** [comment]
>When the tab is clicked, the tab itself is in focus. Probably not a good idea to change focus to the contents because that would interfere with changing tabs with left/right.

I don't know what you mean.

>You would then have to use ctrl+pgdn / ctrl+pgup.

I use ctrl+pgdn / ctrl+pgup now anyway for tab-browsing in the sidebar. It works when I click on the tab-title of the Preview-tab and it works also when the content of the Preview-tab is in focus.

**Comment by xiota** [comment]
Since xitweaks has tab switching detection basically working, I'll take a look at whether the contents can be focused.  However, I'm expecting it will not work for the gtkwebkit widget (used by preview plugin), since it didn't work in the past.

**Comment by ralf3u** [comment]
>[...], I'll take a look at whether the contents can be focused.

This would be great, because it is the last important artifact in the plugin Preview.

**Comment by xiota** [comment]
The xitweaks plugin already has code that moves focus to the contents of the notebooks when sidebar_focus or notebook_focus features are enabled.  It is working on my computer.  Please check whether it is working on yours.

I also found a few minor bugs.  Will try to fix them.

**Comment by ralf3u** [comment]
>Please check whether it is working on yours.

This issue still exists on my computer after `sudo apt update` and `sudo apt dist-ugrade`. If I click on the tab-title on the Preview-tab, the content of the tab is not in focus because I have to press 2x the arrow-down-key of the keyboard, so that a very long webpage starts to scroll downwards.

>It is working on my computer.

Maybe this issue is fixed in Geany 2.0. I use Lubuntu 23.10 and Geany 1.38.

**Comment by xiota** [comment]
Could also be something in gtk3 library.  I remember the signals didn't work a few years ago when you opened these issues.

**Comment by ralf3u** [comment]
In https://github.com/xiota/geany-preview/issues/15#issuecomment-2021850774 you write:

>The xitweaks plugin already has code that moves focus to the contents of the notebooks when sidebar_focus or notebook_focus features are enabled. It is working on my computer. Please check whether it is working on yours.

In Geany in Lubuntu 23.10 the issue is still not fixed. If I click on the Preview-tab-title, the content of the tab is still not in focus.

**Comment by ralf3u** [comment]
>Could also be something in gtk3 library. I remember the signals didn't work a few years ago when you opened these issues.

Could you please retry to solve this issue before I report it to the gtk community.  The are two reasons for that:
- #5 (I opened the issue in 2021) was fixed by you this year thanks to the plugin Xi/Tweaks.
- #11 (I opened the issue in 2021) was fixed in Geany 2.0 that was released in 2023 – I don't think that it was fixed by Geany but by the gtk community.

**Comment by ralf3u** [comment]
>I remember the signals didn't work a few years ago when you opened these issues.

>The xitweaks plugin already has code that moves focus to the contents of the notebooks when sidebar_focus or notebook_focus features are enabled.

At the moment a release of a click on the tab-title puts still the border of the tab in focus instead of the content area of the tab. Please list all signals that you tried, so I can forward them to the gtk community.

**Comment by ralf3u** [comment]
Maybe this information could help you to solve the issue:
If the blinking cursor is in the editor,  and then, if I click in the sidebar on some other tab-title than Preview, like `Documents`, `Symbols` or `Files` (plugin `File Browser`), and then, if I click on the tab-title `Preview`, then the content area of the Preview-tab is in focus.

**Comment by ralf3u** [comment]
Maybe this second information could help you to solve the issue:
If the blinking cursor is in the editor while in the sidebar an other tab-title than `Preview` is active, like `Documents`, `Symbols` or `Files` (plugin `File Browser`), and then, if I click in the sidebar on the tab-title `Preview`, then the content area of the Preview-tab is in focus.

**Comment by ralf3u** [comment]
Maybe this third information could help you to solve the issue:
If the blinking cursor is in the editor while in the sidebar the `Preview`-tab is active, and then, if I click on the tab-title `Preview`, and then, if I click on some other tab-title of the sidebar like `Documents`, `Symbols` or `Files` (plugin `File Browser`), and then, if I click again on the tab-title `Preview`, then the content area of the Preview-tab is **STILL NOT** in focus.

My last two comments and this comment are the evidence for me that gtk3 is not the problem, but the plugin Preview. The plugin `File Browser` works correct, but not the plugin `Preview`.

Maybe it would be good to check which signal the plugin `File Browser` is using to fix this issue.

**Comment by ralf3u** [comment]
An alternative unusual solution how to fix this issue:
Disable all other tab-titles except the tab-title `Preview`. That will make the tab-title `Preview` disappear (see #32), what means that the issue does not exist any more.

**Comment by ralf3u** [comment]
FYI, I don't know where I should mention it, so I mention it here because it is about focus issue: Yesterday I opened a new issue at geany with the issue number 3958. The title of that issue is: `Set the editor in focus when opening a document`. After closing and reopening the issue, the maintainer declared the issue today as bug.
I noticed the issue yesterday. The Preview-tab was in focus while I opened a txt-document from the file-manager.

**Comment by xiota** [comment]
Thanks for letting me know about geany/issue/3958.  I am able to replicate the issue with plugins disabled.

**Comment by ralf3u** [comment]
>My last two comments and this comment are the evidence for me that gtk3 is not the problem, but the plugin Preview. The plugin File Browser works correct, but not the plugin Preview.
Maybe it would be good to check which signal the plugin File Browser is using to fix this issue.

The text above is wrong what I wrote in https://github.com/xiota/geany-preview/issues/15#issuecomment-2351783068 because I can reproduce the issue with the tab `Documents` while all plugins are disabled. I will report the issue at geany.

Because the issue has to be fixed upstream, I close this issue.

**Comment by ralf3u** [comment]
I just opened a new issue at geany with the number 3961. The title of the issue is:
`BUG: A release of a click on a tab-title of the sidebar does not put always the content-area of the tab in focus`.

**Comment by ralf3u** [comment]
The maintainer of Geany wrote some hours ago this text:
>Also don't know what the GTK4 behaviour is, maybe find a simple GTK4 app and see if its the same?

GTK4 was released in 2020. I don't know a simple GTK4 app for testing if the issue is still there. Do you know a simple GTK4 app for testing?

**Comment by ralf3u** [comment]
>Do you know a simple GTK4 app for testing?

I just recognized that the question makes no sense, because even if a simple GTK4 app would exist and even if in the simple GTK4 app the issue would not exist any more, it does not mean that the simple GTK4 app is using the GTK4 default behavior. As with Geany, there may be a non-standard behaviour implemented.

**Comment by xiota** [comment]
Geany devs would know better than I do.  You'd probably have to write the "simple" app yourself to find the default behavior (for reason you already noticed).  The result probably wouldn't affect Geany for years.

**Comment by ralf3u** [comment]
Because the issue that I mentioned in comment https://github.com/xiota/geany-preview/issues/15#issuecomment-2372261705 was declared as invalid, I just opened at geany one further issue with the issue number 3966. The title of the issue is:
`Feature request: When clicking on a tab-title of the sidebar, the focus should always be set on the content-area of the tab`

**Comment by ralf3u** [comment]
Today I wrote at geany with the issue number 3961 that I can't confirm that sidebar and message window have the same behavior. Later, when I deactivated the plugins, I recognized that I was wrong. The plugin Xi/Tweaks was the cause.
The plugin Xi/Tweaks is a big help because thanks to the plugin I can see on an eink-display in text-mode which tab has focus, editor, sidebar or message window. But the plugin is also changing the behavior of the arrow-keys in  the message window :
- If the tab has focus, so not the content-area of the tab, then tab-browsing with an arrow-key is not possible.
- If the tab has focus, so not the content-area of the tab, I can not put the content-area of the tab in focus by pressing 1x an arrow-key.

Maybe it would be good if both default things would work when the plugin Xi/Tweaks is activated.

**Comment by xiota** [comment]
Are you requesting... that xitweaks be modified... to move focus to sidebar contents when sidebar tabs clicked... because that is what it appears to already do for the msgwin?

If true that it already changes focus for msgwin, I can look into implementing similar for sidebar.

**Comment by ralf3u** [comment]
>Are you requesting... that xitweaks be modified... to move focus to sidebar contents when sidebar tabs clicked... because that is what it appears to already do for the msgwin?

I can't confirm that focus on the content-area of msgwin is working:
- Open geany, click on the tab-title Terminal: content-area of the tab is in focus.
- Open geany, click on the tab-title Status and then on the tab-title Terminal: content-area of the tab is not in focus.
- Open geany, click on the tab-title Status: content-area of the tab is not in focus.

What I wanted to say with my last comment one hour ago to the subject msgwin:
The plugin Xi/Tweaks seems to disable the possibility to use the arrow-keys for tab-browsing and for putting the content-area in focus. It would be better if Xi/Tweaks would not disable it.

**Comment by xiota** [comment]
> The plugin Xi/Tweaks seems to disable the possibility to use the arrow-keys for tab-browsing and for putting the content-area in focus. It would be better if Xi/Tweaks would not disable it.

I don't think that is caused by the xitweaks plugin, but will check.  With all plugins disabled, depending on which msgwin tab I click, focus moves to contents.  Since focus moved, the tabs no longer receive keyboard input, so arrow keys cannot switch them.

**Comment by ralf3u** [comment]
All code in the plugin `Preview` should be removed that is trying to put the focus on the content-area of a tab when the tab-title was clicked.

**Comment by xiota** [comment]
> All code in the plugin Preview should be removed that is trying to put the focus on the content-area of a tab when the tab-title was clicked.

I have no problem removing code that attempts to change focus.  Note that any such code was added by your previous request and, after removal, would not be added back in future.

**Comment by ralf3u** [comment]
>I have no problem removing code that attempts to change focus. Note that any such code was added by your previous request and, after removal, would not be added back in future.

The problem is that it didn't work. If you find a way to make it work, then of course I'm not for a removal.

**Comment by ralf3u** [comment]
>Note that any such code was added by your previous request and, after removal, would not be added back in future.

Since 2021 the code does not work, and it can be assumed that it will never work – at least not with this code.

**Comment by ralf3u** [comment]
FYI, I just opened a new issue at gtk:
https://gitlab.gnome.org/GNOME/gtk/-/issues/7054
The title of the issue is:
`When clicking on a tab-title, the focus should ALWAYS be set on the content-area of the tab`

**Comment by xiota** [comment]
> can be assumed that it will never work – at least with this code

I've pretty much given up on trying.

> opened a new issue at gtk

I suspect they will say something like... that's application developer responsibility.

**Comment by ralf3u** [comment]
Here are three possibilities for users who don't want to get touched with the issue:
- don't click on the tab-title; click on the content-area of the tab instead
- use the shortcut `Toggle focus between editor and sidebar`
- except the Preview-tab, disable all other tabs (that will disable the tab-bar)

**Comment by ralf3u** [comment]
>I've pretty much given up on trying.

Just to be on the safe side:
Did you remove all code of the plugin `Preview` that tries to put the focus on the content-area of the tab when the tab-title was clicked?

**Comment by xiota** [comment]
I haven't made any changea yet.

**Comment by ralf3u** [comment]
FYI, I still didn't get an answer from gtk. I removed the following part of my report at gtk, because maybe I overstrained the developers with my questions:

**Three questions**
1. Do you intend to implement the feature (see title of the issue) as default behavior in GTK3?
2. Is the feature (see title of the issue) already default behavior in GTK4?
3. If the feature (see title of the issue) is not default behavior in GTK4, do you intend to implement the feature as default behavior in GTK4?

The fact is that they didn't close the issue. So, maybe they are aware of the issue. Maybe they know, that if the issue will be fixed, it will requires too many further changes, so that nobody wants to get in touch with that issue.
It seems to be that the only way to fix the issue is the issue number 3966 at geany. But it is not so dramatic if the issue will not be fixed at geany because I mentioned some possibilities in https://github.com/xiota/geany-preview/issues/15#issuecomment-2392431389 how to avoid the contact with the issue.

**Comment by ralf3u** [comment]
FYI, I still didn't get an answer from gtk. I removed the following part of my report at gtk, because maybe questions are not welcome:

**Three questions**
1. Do you intend to implement the feature (see title of the issue) as default behavior in GTK3?
2. Is the feature (see title of the issue) already default behavior in GTK4?
3. If the feature (see title of the issue) is not default behavior in GTK4, do you intend to implement the feature as default behavior in GTK4?

The fact is that they didn't close the issue.

It seems to be that the only way to fix the issue in the plugin `Preview` is the issue number 3966 at geany. But it is not so dramatic if the issue will not be fixed at geany because I mentioned some possibilities in https://github.com/xiota/geany-preview/issues/15#issuecomment-2392431389 how to avoid the contact with the issue in the plugin `Preview`.

**Comment by ralf3u** [comment]
FYI, the chief maintainer of Geany wrote yesterday some sentences to the issue 3966.

**Comment by ralf3u** [comment]
>I haven't made any changes yet.

It could be that the non-working focus-code is causing the grapical glitch from issue number 61, like the non-working focus-code of the plugin `Xi/Tweaks` that causes that a navigation of the arrow-keys is not possible (issue number 58).

It could be that the non-working focus-code is causing the focus-code of the `geanylua`-`switch-views`-shortcuts does not work, like the non-working focus-code of the plugin `Xi/Tweaks` that causes that a navigation of the arrow-keys is not possible (issue number 58).

It could be that the non-working focus-code is causing that the Geany shortcut `Toggle Sidebar` does not work correctly with the `geanylua`-`switch-views`-shortcuts, like the non-working focus-code of the plugin `Xi/Tweaks` that causes that a navigation of the arrow-keys is not possible (issue number 58).

Please remove entirely the focus-code of the plugin `Preview`. I suggest that you remove also entirely the shortcut `Toggle focus between editor and sidebar` of the plugin `Preview` (and to paste that shortcut-code in a TXT-document as backup), so that there will not even be a small piece of code that is organising the focus.

Please tell me when you removed the non-working focus-code and the shortcut `Toggle focus between editor and sidebar` from the plugin `Preview`, so I will update my system.

Then I can continue my tests to see if the grapical glitch is still there, if the focus of `geanylua`-`switch-views`-shortcuts works correctly and if the Geany shortcut `Toggle Sidebar`works properly with the `geanylua`-`switch-views`-shortcuts.
